File: backend/gas_station.ts
/**
 * gas_station.ts — Gas Balance Monitor for Valentin
 *
 * Checks agent and admin SUI balances and warns when gas is low.
 * Can run standalone or be imported into the relayer.
 *
 * Usage:
 *   npx ts-node gas_station.ts              # one-shot check
 *   npx ts-node gas_station.ts --watch      # continuous monitoring
 *   # or: npm run gas
 */

import {
  SuiClient,
  getFullnodeUrl,
} from '@mysten/sui/client';
import { Ed25519Keypair } from '@mysten/sui/keypairs/ed25519';
import { fromBase64 } from '@mysten/sui/utils';
import 'dotenv/config';

// ═══════════════════════════════════════════════════════════
//  CONFIG
// ═══════════════════════════════════════════════════════════

const NETWORK = (process.env.SUI_NETWORK as 'devnet' | 'testnet' | 'mainnet') ?? 'devnet';
const RPC_URL = process.env.SUI_RPC_URL ?? getFullnodeUrl(NETWORK);
const client  = new SuiClient({ url: RPC_URL });

/** Minimum gas balance before we fire a warning (in MIST) */
const MIN_GAS_MIST       = BigInt(process.env.MIN_GAS_MIST ?? '500000000');     // 0.5 SUI
const CRITICAL_GAS_MIST  = BigInt(process.env.CRITICAL_GAS_MIST ?? '100000000'); // 0.1 SUI
const FAUCET_URLS: Record<string, string> = {
  devnet:  'https://faucet.devnet.sui.io/gas',
  testnet: 'https://faucet.testnet.sui.io/v1/gas',
};

// ═══════════════════════════════════════════════════════════
//  HELPERS
// ═══════════════════════════════════════════════════════════

function loadAddress(envKey: string): string | null {
  const key = process.env[envKey];
  if (!key) return null;

  try {
    if (key.startsWith('suiprivkey')) {
      return Ed25519Keypair.fromSecretKey(key).getPublicKey().toSuiAddress();
    }
    const bytes = key.startsWith('0x')
      ? Uint8Array.from(Buffer.from(key.slice(2), 'hex'))
      : fromBase64(key);
    return Ed25519Keypair.fromSecretKey(bytes).getPublicKey().toSuiAddress();
  } catch {
    return null;
  }
}

export interface GasStatus {
  address: string;
  role: string;
  balanceMist: bigint;
  balanceSui: string;
  level: 'ok' | 'low' | 'critical' | 'empty';
  message: string;
}

async function checkBalance(address: string, role: string): Promise<GasStatus> {
  const coins = await client.getCoins({
    owner: address,
    coinType: '0x2::sui::SUI',
  });

  const total = coins.data.reduce(
    (acc: bigint, c: any) => acc + BigInt(c.balance),
    BigInt(0),
  );

  let level: GasStatus['level'];
  let message: string;

  if (total === BigInt(0)) {
    level = 'empty';
    message = `🚨 ${role} hat KEIN Gas! Sofort Faucet nutzen.`;
  } else if (total < CRITICAL_GAS_MIST) {
    level = 'critical';
    message = `🔴 ${role} Gas KRITISCH: ${formatSui(total)} SUI — Trades werden bald fehlschlagen!`;
  } else if (total < MIN_GAS_MIST) {
    level = 'low';
    message = `🟡 ${role} Gas niedrig: ${formatSui(total)} SUI — bald auffüllen.`;
  } else {
    level = 'ok';
    message = `🟢 ${role} Gas OK: ${formatSui(total)} SUI`;
  }

  return {
    address,
    role,
    balanceMist: total,
    balanceSui: formatSui(total),
    level,
    message,
  };
}

function formatSui(mist: bigint): string {
  return (Number(mist) / 1e9).toFixed(4);
}

// ═══════════════════════════════════════════════════════════
//  PUBLIC API (importable by relayer)
// ═══════════════════════════════════════════════════════════

export async function checkAllGas(): Promise<GasStatus[]> {
  const results: GasStatus[] = [];

  const agentAddr = loadAddress('AGENT_PRIVATE_KEY');
  const adminAddr = loadAddress('ADMIN_PRIVATE_KEY');

  if (agentAddr) {
    results.push(await checkBalance(agentAddr, 'Agent (Valentin)'));
  }
  if (adminAddr) {
    results.push(await checkBalance(adminAddr, 'Admin (Korbinian)'));
  }

  return results;
}

export async function isGasSufficient(): Promise<boolean> {
  const statuses = await checkAllGas();
  return statuses.every((s) => s.level === 'ok' || s.level === 'low');
}

// ═══════════════════════════════════════════════════════════
//  AUTO-FAUCET (devnet/testnet only)
// ═══════════════════════════════════════════════════════════

async function requestFaucet(address: string): Promise<boolean> {
  const url = FAUCET_URLS[NETWORK];
  if (!url) {
    console.log(`   ⚠️  Kein Faucet für ${NETWORK} verfügbar.`);
    return false;
  }

  try {
    console.log(`   💧 Faucet-Anfrage für ${address}...`);
    const resp = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        FixedAmountRequest: { recipient: address },
      }),
    });

    if (resp.ok) {
      console.log(`   ✅ Faucet erfolgreich!`);
      return true;
    } else {
      console.log(`   ❌ Faucet-Fehler: ${resp.status} ${resp.statusText}`);
      return false;
    }
  } catch (e: any) {
    console.log(`   ❌ Faucet-Fehler: ${e.message}`);
    return false;
  }
}

// ═══════════════════════════════════════════════════════════
//  CLI
// ═══════════════════════════════════════════════════════════

async function printStatus() {
  console.log('\n═══════════════════════════════════════════════════');
  console.log('  ⛽ Gas Station — Balance Monitor');
  console.log('═══════════════════════════════════════════════════');
  console.log(`  Network:  ${NETWORK}`);
  console.log(`  Min Gas:  ${formatSui(MIN_GAS_MIST)} SUI`);
  console.log(`  Critical: ${formatSui(CRITICAL_GAS_MIST)} SUI\n`);

  const statuses = await checkAllGas();

  if (statuses.length === 0) {
    console.log('  ⚠️  Keine Keys in .env gefunden (AGENT_PRIVATE_KEY / ADMIN_PRIVATE_KEY)');
    return statuses;
  }

  for (const s of statuses) {
    console.log(`  ${s.message}`);
    console.log(`     Address: ${s.address}`);
    console.log(`     Balance: ${s.balanceMist.toString()} MIST\n`);

    // Auto-faucet if critical/empty on devnet/testnet
    if ((s.level === 'critical' || s.level === 'empty') && FAUCET_URLS[NETWORK]) {
      await requestFaucet(s.address);
    }
  }

  return statuses;
}

async function main() {
  const watchMode = process.argv.includes('--watch');

  await printStatus();

  if (watchMode) {
    const interval = parseInt(process.env.GAS_CHECK_INTERVAL ?? '30000', 10);
    console.log(`\n  👀 Watch-Modus: Prüfe alle ${interval / 1000}s ...\n`);

    setInterval(async () => {
      const statuses = await checkAllGas();
      for (const s of statuses) {
        if (s.level !== 'ok') {
          console.log(`  ${s.message}`);
          if ((s.level === 'critical' || s.level === 'empty') && FAUCET_URLS[NETWORK]) {
            await requestFaucet(s.address);
          }
        }
      }
    }, interval);
  }
}

main().catch(console.error);

File: backend/blockchain/gas_station.py
#!/usr/bin/env python3
"""
gas_station.py — Gas Balance Monitor

Checks agent and admin SUI balances and warns when gas is low.
Can run standalone or be imported into the relayer.

Usage:
    python gas_station.py              # one-shot check
    python gas_station.py --watch      # continuous monitoring
"""

import asyncio
import logging
import os
import sys
from dataclasses import dataclass
from typing import List, Optional

import httpx
from dotenv import load_dotenv

load_dotenv()

logger = logging.getLogger(__name__)

# ═══════════════════════════════════════════════════════════
#  CONFIG
# ═══════════════════════════════════════════════════════════

NETWORK = os.getenv("SUI_NETWORK", "devnet")
RPC_URL = os.getenv("SUI_RPC_URL", f"https://fullnode.{NETWORK}.sui.io:443")

MIN_GAS_MIST = int(os.getenv("MIN_GAS_MIST", "500000000"))  # 0.5 SUI
CRITICAL_GAS_MIST = int(os.getenv("CRITICAL_GAS_MIST", "100000000"))  # 0.1 SUI

FAUCET_URLS = {
    "devnet": "https://faucet.devnet.sui.io/gas",
    "testnet": "https://faucet.testnet.sui.io/v1/gas",
}

# ═══════════════════════════════════════════════════════════
#  HELPERS
# ═══════════════════════════════════════════════════════════

_rpc_id = 0


def _rpc_call(method: str, params: list) -> dict:
    """Minimal Sui JSON-RPC call."""
    global _rpc_id
    _rpc_id += 1
    payload = {
        "jsonrpc": "2.0",
        "id": _rpc_id,
        "method": method,
        "params": params,
    }
    with httpx.Client(timeout=15) as client:
        resp = client.post(RPC_URL, json=payload)
        resp.raise_for_status()
        data = resp.json()
    if "error" in data:
        raise RuntimeError(f"RPC error: {data['error']}")
    return data.get("result", {})


def _load_address(env_key: str) -> Optional[str]:
    """Derive a Sui address from a private key env var using the sui CLI."""
    key = os.getenv(env_key)
    if not key:
        return None

    try:
        import subprocess

        # Use sui keytool to convert private key to address
        result = subprocess.run(
            ["sui", "keytool", "show", key],
            capture_output=True,
            text=True,
            timeout=10,
        )
        if result.returncode == 0:
            # Parse address from output
            for line in result.stdout.splitlines():
                line = line.strip()
                if line.startswith("Sui Address:") or line.startswith("suiAddress:"):
                    return line.split(":", 1)[1].strip()
                # Also try JSON output
                if '"suiAddress"' in line or '"sui_address"' in line:
                    import json

                    data = json.loads(result.stdout)
                    return data.get("suiAddress", data.get("sui_address", ""))
            # Try parsing as JSON
            import json

            try:
                data = json.loads(result.stdout)
                return data.get("suiAddress", data.get("sui_address", ""))
            except json.JSONDecodeError:
                pass
    except (subprocess.TimeoutExpired, FileNotFoundError):
        pass

    # Fallback: if key looks like an address, use it directly
    if key.startswith("0x") and len(key) >= 42:
        return key

    return None


@dataclass
class GasStatus:
    address: str
    role: str
    balance_mist: int
    balance_sui: str
    level: str  # 'ok' | 'low' | 'critical' | 'empty'
    message: str


def _format_sui(mist: int) -> str:
    return f"{mist / 1e9:.4f}"


def _check_balance(address: str, role: str) -> GasStatus:
    """Check gas balance for a single address."""
    result = _rpc_call("suix_getCoins", [address, "0x2::sui::SUI", None, None])
    coins = result.get("data", [])
    total = sum(int(c.get("balance", "0")) for c in coins)

    if total == 0:
        level = "empty"
        message = f" {role} hat KEIN Gas! Sofort Faucet nutzen."
    elif total < CRITICAL_GAS_MIST:
        level = "critical"
        message = (
            f" {role} Gas KRITISCH: {_format_sui(total)} SUI — Trades werden bald fehlschlagen!"
        )
    elif total < MIN_GAS_MIST:
        level = "low"
        message = f"🟡 {role} Gas niedrig: {_format_sui(total)} SUI — bald auffüllen."
    else:
        level = "ok"
        message = f"🟢 {role} Gas OK: {_format_sui(total)} SUI"

    return GasStatus(
        address=address,
        role=role,
        balance_mist=total,
        balance_sui=_format_sui(total),
        level=level,
        message=message,
    )


# ═══════════════════════════════════════════════════════════
#  PUBLIC API (importable by relayer)
# ═══════════════════════════════════════════════════════════


def check_all_gas() -> List[GasStatus]:
    """Check gas balances for all configured accounts."""
    results = []

    agent_addr = _load_address("AGENT_PRIVATE_KEY")
    admin_addr = _load_address("ADMIN_PRIVATE_KEY")

    if agent_addr:
        results.append(_check_balance(agent_addr, "Agent (Valentin)"))
    if admin_addr:
        results.append(_check_balance(admin_addr, "Admin (Korbinian)"))

    return results


def is_gas_sufficient() -> bool:
    """True if all accounts have at least 'low' gas."""
    statuses = check_all_gas()
    return all(s.level in ("ok", "low") for s in statuses)


# ═══════════════════════════════════════════════════════════
#  AUTO-FAUCET (devnet/testnet only)
# ═══════════════════════════════════════════════════════════


def _request_faucet(address: str) -> bool:
    url = FAUCET_URLS.get(NETWORK)
    if not url:
        print(f"     Kein Faucet für {NETWORK} verfügbar.")
        return False

    try:
        print(f"    Faucet-Anfrage für {address}...")
        with httpx.Client(timeout=30) as client:
            resp = client.post(
                url,
                json={"FixedAmountRequest": {"recipient": address}},
            )
        if resp.is_success:
            print("    Faucet erfolgreich!")
            return True
        else:
            print(f"    Faucet-Fehler: {resp.status_code} {resp.reason_phrase}")
            return False
    except Exception as e:
        print(f"    Faucet-Fehler: {e}")
        return False


# ═══════════════════════════════════════════════════════════
#  CLI
# ═══════════════════════════════════════════════════════════


def _print_status() -> List[GasStatus]:
    print("\n═══════════════════════════════════════════════════")
    print("   Gas Station — Balance Monitor")
    print("═══════════════════════════════════════════════════")
    print(f"  Network:  {NETWORK}")
    print(f"  Min Gas:  {_format_sui(MIN_GAS_MIST)} SUI")
    print(f"  Critical: {_format_sui(CRITICAL_GAS_MIST)} SUI\n")

    statuses = check_all_gas()

    if not statuses:
        print("    Keine Keys in .env gefunden (AGENT_PRIVATE_KEY / ADMIN_PRIVATE_KEY)")
        return statuses

    for s in statuses:
        print(f"  {s.message}")
        print(f"     Address: {s.address}")
        print(f"     Balance: {s.balance_mist} MIST\n")

        # Auto-faucet if critical/empty on devnet/testnet
        if s.level in ("critical", "empty") and NETWORK in FAUCET_URLS:
            _request_faucet(s.address)

    return statuses


async def _watch_loop():
    interval = int(os.getenv("GAS_CHECK_INTERVAL", "30000")) / 1000
    print(f"\n   Watch-Modus: Prüfe alle {interval:.0f}s ...\n")

    while True:
        await asyncio.sleep(interval)
        statuses = check_all_gas()
        for s in statuses:
            if s.level != "ok":
                print(f"  {s.message}")
                if s.level in ("critical", "empty") and NETWORK in FAUCET_URLS:
                    _request_faucet(s.address)


def main():
    _print_status()

    if "--watch" in sys.argv:
        asyncio.run(_watch_loop())


if __name__ == "__main__":
    main()
